### DISCUSSION TITLE
i18n and templates

### DISCUSSION THREAD

**Discussion by lamyseba** [jeremyevans/rodauth#202] [General]
Hi,
First of all thanks for having done and published Rodauth
I'm using it in a french website, and I would like to override the default templates, especially the one used to send a "verify email" or "reset password_email". I know I can override the "create_verify_account_email" or "send_verify_account_email" methods, but I think there will be other use-case where I want to translate a text hardcoded in english in the templates. Don't know how I should do that and dit not find anything in the docs, except in base.rdoc:

> cache_templates | Whether to cache templates. True by default. It may be worth  switching this to false in development if you are using your own  templates instead of the templates provided by Rodauth.

This means it should be possible, but I found no guide to do this. Where should I put my template files? Where should I declare that I want to override defaults template with my template files?  Maybe I could override every method using default templates, but this seems not to be an easy/standard way.

Also, in the external-feature, the rodauth-i18n seems not care about these templates. This means eg. verification email is sent with an english text, even in a translated app ? 
I think a clean way to translate a template inside rodauth-i18n would be the best solution to my problem. 

Thanks for your help.

**Comment by jeremyevans** [comment]
I don't think `cache_templates` will matter in terms of i18n.  You should use the existing configuration methods to change the emails that are sent.   I'm guessing most applications override the email methods even if they are in English, because the default emails are very simple. I think rodauth-i18n takes care of most or all of the non-email related translation work.  It doesn't currently have a French translation, but I'm guessing they would accept one via a pull request.  You might want to ask the rodauth-18n developers what they think about translated emails.

**Comment by janko** [comment]
I'm not sure what rodauth-i18n could do about translating email templates. If a Rodauth method doesn't wrap the returned text with a `translate` call, the alternative that rodauth-i18n takes is to re-define that method, and have it go through `I18n.translate`.

But in order for this to work, the method needs to be defined on the Rodauth object before the `i18n` feature is loaded. So far, the only Rodauth methods that aren't using `#translate` are certain validation error messages in `login_and_password_requirements_base`. So, to ensure those methods are loaded beforehand, the `i18n` feature declares `login_and_password_requirements_base` as a dependency. It's a small trade-off considering that this feature will probably get loaded one way or the other in most applications.

Obviously, we couldn't scale this approach to email body methods, because that would require `i18n` feature to depend on all features that defines those methods. However, instead of going through the I18n gem, perhaps I could add the ability for people to submit localized email templates to the rodauth-i18n repository. That could work 🤔

You request is mainly about wanting to have email templates translated by default, correct?

**Reply by lamyseba** [reply]
Yes, my request is mainly about the ability to translate email templates. I don't know what you are meaning with "By default". Ideally yes, if you set a language, all rodauth should be available in that language, even email sent to users.
I think the ability to submit localized email template is the good way to go. But, as I said in my first message,  I don't even know how to override a default template and have Rodauth set to use my localized template. 

* Used email_templates_path are not in the the Auth Value Methods of Rodauth plugins.
* I found no simple way to have a template directory in my project whose files (if they are present) are used instead of the default

So the only way I found is to override the Auth Methods that use the english email templates, but that seemed to me quite a dirty way to go for translation.

**Reply by janko** [reply]
The default directory for Roda's [`render`](https://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/Render.html) plugin (which Rodauth uses) is `views/` in your project root. So, if you create templates in that directory with the same name as Rodauth's built-in templates, Rodauth will automatically pick those instead.

See the example below which shows the custom email verification template being picked up:

```rb
require "rodauth"
require "sequel"
require "mail"

DB = Sequel.sqlite
DB.create_table :accounts do
  primary_key :id
  String :email, null: false
  String :status_id, null: false
  String :password_hash
end
DB.create_table :account_verification_keys do
  foreign_key :id, :accounts, primary_key: true
  String :key, null: false
  DateTime :requested_at, null: false, default: Sequel::CURRENT_TIMESTAMP
  DateTime :email_last_sent, null: false, default: Sequel::CURRENT_TIMESTAMP
end

Mail.defaults do
  delivery_method :test
end

rodauth = Rodauth.lib do
  enable :create_account, :verify_account
  account_password_hash_column :password_hash
  domain "example.com"
end

rodauth.create_account(login: "user@example.com", password: "secret")
email = Mail::TestMailer.deliveries.last
email.body #=> "My custom template for account verification. Follow this link: https://example.com/verify-account?key=1_BTVLpBYnW7TOAYH_uqpE0da7azocfusWjTEbZRVuExY"
```
```erb
<!-- views/verify-account-email.erb -->
My custom template for account verification. Follow this link: <%= rodauth.verify_account_email_link %>
```

Does that help?

**Comment by lamyseba** [comment]
Yes, that help a lot! Thank you!
